Add experimental Span Suggesters

thomashacker opened the pull request:
This `PR` adds three new experimental suggester functions for the `spancat` component and a spaCy project showcasing how to use them in a `config.cfg` file.

#### Subtree Suggester:
- Uses annotations from the Tagger and Parser to suggests `subtrees` of individual tokens

#### Chunk Suggester:
- Uses annotations from the Tagger and Parser to suggest `noun_chunks`

#### Sentence Suggester:
- Uses sentence boundaries to suggest `sentences`

These suggesters also come with the `ngram` functionality which allows users to set a list of `sizes` for suggesting individual ngrams

#### The spaCy project covers:
- How to source components from existing models
- How to use `frozen_components` & `annotating_components`
- How to use custom suggester functions registered in the `registry`

adrianeboyd commented on projects/span_suggesters/configs/chunk_suggester.cfg:
```diff
@@ -0,0 +1,151 @@
+dev = null
+vectors = null
+init_tok2vec = null
+span_key = null
```
I think this should be consistently called `spans_key` and it shouldn't be under the `[paths]` section because that will be confusing for users. Even if it's shared between multiple components, you can define it in one spot (in spancat?) and then refer to that config variable in any other section, just like we do for tok2vec widths in the default configs:

```ini[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3
```

adrianeboyd commented:
The general plan is to initially release things like this under `spacy-experimental` for easier testing, and then move features to the core library or other places once they seem stable and generally useful. Not everything will make the cut and the APIs might change a bit, but I think these particular suggesters have a very high chance of being moved to the core library soon.

adrianeboyd commented:
I think that these span suggesters would be simpler and faster without the ngrams thrown in, and it would make sense to have a more general way of combining arbitrary suggesters/suggestions than what's proposed here. (In particular sentences + ngrams doesn't make a lot of sense to me?)

thomashacker commented:
I think, having the ngram functionality as a baseline will be useful for most use-cases. For example, the `sentences + ngrams` could be used for the healthsea dataset. I also like that it's optional to use them, when leaving the sizes list empty, the suggester will skip the whole code block that handles the ngram suggesting.

adrianeboyd commented:
It is a problem that it's hard to specify a variable number of things (each with their own configs) to combine in a config block, which is something that's also come up with augmenters.

If we're going to do this, I think it would be better to have a utility function that merges suggestions in a more general, efficient way. The `ops.to_numpy()` for each row in the array is going to be particularly slow in the current version and appending everything row-by-row vs. working with the existing numpy array looks like it may also slow things down. (In a case where you already know you want it as numpy, I think you could just pass `NumpyOps` as the ops instead to the suggester and then you could skip all the conversions?)

I decided to sketch out what a suggestions merger could look like (could still use some typing and testing, particularly on GPU; note that there's no `cupy.unique` so you have to convert to `numpy`):

```python
from typing import List, Optional, Iterable, cast
import numpy
from thinc.api import get_current_ops, Ops
from thinc.types import Ragged, Ints1d
from spacy.tokens import Doc
from spacy.util import registry
from spacy.pipeline.spancat import Suggester


@registry.misc("experimental.ngram_sentence_suggester.v1")
def build_ngram_sentence_suggester(sizes: List[int]) -> Suggester:
    """Suggest ngrams and sentences. Requires sentence boundaries"""

    ngram_suggester = registry.misc.get("spacy.ngram_suggester.v1")(sizes)
    def ngram_sentence_suggester(
        docs: Iterable[Doc], *, ops: Optional[Ops] = None
    ) -> Ragged:
        ngram_suggestions = ngram_suggester(docs, ops=ops)
        sentence_suggestions = sentence_suggester(docs, ops=ops)
        return merge_suggestions([ngram_suggestions, sentence_suggestions], ops=ops)

    return sentence_suggester


def sentence_suggester(docs: Iterable[Doc], *, ops: Optional[Ops] = None) -> Ragged:
    if ops is None:
        ops = get_current_ops()
    spans = []
    lengths = []

    for doc in docs:
        sents = list(doc.sents)
        spans.extend((sent.start, sent.end) for sent in sents)
        lengths.append(len(sents))

    lengths_array = cast(Ints1d, ops.asarray(lengths, dtype="i"))
    if len(spans) > 0:
        output = Ragged(ops.asarray(spans, dtype="i"), lengths_array)
    else:
        output = Ragged(ops.xp.zeros((0, 0), dtype="i"), lengths_array)

    return output


def merge_suggestions(suggestions: List[Ragged], ops: Optional[Ops] = None) -> Ragged:
    if ops is None:
        ops = get_current_ops()

    spans = []
    lengths = []

    if len(suggestions) == 0:
        lengths_array = cast(Ints1d, ops.asarray(lengths, dtype="i"))
        return Ragged(ops.xp.zeros((0, 0), dtype="i"), lengths_array)

    len_docs = len(suggestions[0])
    assert all(len_docs == len(x) for x in suggestions)

    for i in range(len_docs):
        combined = ops.xp.vstack([s[i].data for s in suggestions])
        uniqued = numpy.unique(ops.to_numpy(combined), axis=0)
        spans.append(ops.asarray(uniqued))
        lengths.append(uniqued.shape[0])

    lengths_array = cast(Ints1d, ops.asarray(lengths, dtype="i"))
    if len(spans) > 0:
        output = Ragged(ops.xp.vstack(spans), lengths_array)
    else:
        output = Ragged(ops.xp.zeros((0, 0), dtype="i"), lengths_array)

    return output
```

And I wouldn't be surprised if there's a faster way to manage the per-doc unique step, but this has reached the end of my numpy knowledge.

thomashacker commented:
Wow, that's awesome! Thanks adriane! 🥳 
I've implemented the `merge_suggesters` function and also added suggester registries that don't use the ngram at all, if the user does not want to:

```
  spacy-experimental.subtree_suggester.v1
  spacy-experimental.ngram_subtree_suggester.v1 
  spacy-experimental.chunk_suggester.v1 
  spacy-experimental.ngram_chunk_suggester.v1 
  spacy-experimental.sentence_suggester.v1 
  spacy-experimental.ngram_sentence_suggester.v1 
```

adrianeboyd commented:
Can you remove the formatting changes to `azure-pipelines.yml`?

adrianeboyd commented on projects/span_suggesters/configs/sentence_suggester.cfg:
```diff
@@ -0,0 +1,150 @@
+max_epochs = 0
+max_steps = 20000
+eval_frequency = 200
+frozen_components = ["tok2vec","parser","tagger"]
+annotating_components = ["tagger","parser"]
```
This only needs the parser, I would remove the tagger entirely to make this clear.

adrianeboyd replied:
You could also use a sentencizer without any source pipeline as long as its annotating.

adrianeboyd commented on projects/span_suggesters/configs/chunk_suggester.cfg:
```diff
@@ -0,0 +1,150 @@
+max_epochs = 0
+max_steps = 20000
+eval_frequency = 200
+frozen_components = ["tok2vec","tagger","parser"]
+annotating_components = ["tagger","parser"]
```
This is more complicated and depends on the source pipeline. In this case you need: tagger, parser, attribute_ruler because the noun chunk requires POS and deps.

adrianeboyd commented on projects/span_suggesters/configs/subtree_suggester.cfg:
```diff
@@ -0,0 +1,149 @@
+max_epochs = 0
+max_steps = 20000
+eval_frequency = 200
+frozen_components = ["tok2vec","parser","tagger"]
+annotating_components = ["tagger","parser"]
```
I would completely remove the tagger here, too.

adrianeboyd commented on spacy_experimental/span_suggesters/tests/test__suggesters.py:
```diff
@@ -0,0 +1,60 @@
+from spacy.util import registry
```
Update the file name to `test_suggesters.py`

svlandeg commented on spacy_experimental/span_suggesters/chunk_suggester.py:
```diff
@@ -0,0 +1,55 @@
+from .merge_suggesters import merge_suggestions
+
+
+@registry.misc("experimental.ngram_chunk_suggester.v1")
```
Because we have them as entry points in `setup.cfg`, we can remove the decorators from these functions (also ensure that the names don't go out-of-sync, which they have here)

adrianeboyd commented on setup.cfg:
```diff
@@ -30,6 +30,15 @@ spacy_scorers =
     spacy-experimental.tokenizer_senter_scorer.v1 = spacy_experimental.char_tokenizer.scorers:make_tokenizer_senter_scorer
 spacy_tokenizers =
     spacy-experimental.char_pretokenizer.v1 = spacy_experimental.char_tokenizer.char_pretokenizer:char_pretokenizer_v1
+spacy_misc =
+    spacy-experimental.subtree_suggester.v1 = spacy_experimental.span_suggesters.subtree_suggester:build_subtree_suggester
+    spacy-experimental.ngram_subtree_suggester.v1 = spacy_experimental.span_suggesters.subtree_suggester:build_ngram_subtree_suggester
+    spacy-experimental.chunk_suggester.v1 = spacy_experimental.span_suggesters.chunk_suggester:build_chunk_suggester
+    spacy-experimental.ngram_chunk_suggester.v1 = spacy_experimental.span_suggesters.chunk_suggester:build_ngram_chunk_suggester
+    spacy-experimental.sentence_suggester.v1 = spacy_experimental.span_suggesters.sentence_suggester:build_sentence_suggester
+    spacy-experimental.ngram_sentence_suggester.v1 = spacy_experimental.span_suggesters.sentence_suggester:build_ngram_sentence_suggester
```
Since this is experimental, my feeling is more that it's okay to provide a wider/noisy range of things to experiment with easily, but we wouldn't necessarily do the same in spacy later.

thomashacker replied:
I think it's good to have at least some of these as an example of how it's done? I also agree that we should change it when migrating them to spaCy later.

svlandeg replied:
Ok, fair enough!

adrianeboyd commented:
Can you add documentation for these to the top-level README?

adrianeboyd commented:
Can you add sample configs for one of the types of suggesters that also explains how to use `sizes` for the `ngram` versions?

Lolologist commented:
Not to keep nosing into y'all's business, but once this merges would consumer (my) feedback on trying a model using this be warranted/appreciated?

adrianeboyd commented:
@Lolologist: Feedback is welcome, that's pretty much what `spacy-experimental` is there for! We want it to be easy for users to try things out so we can improve them before they're added to the core library.

thomashacker commented:
Added example snippets to the readme with [1f52f1e](https://github.com/explosion/spacy-experimental/pull/11/commits/1f52f1e6c9ea1a0248e1a05584fa232df6762da1) 🎉

adrianeboyd commented on README.md:
````diff
@@ -249,6 +249,41 @@ None currently.
+sizes = [1, 2, 3]
+```
+
+> Note that all the custom suggester functions are registered in `@misc`
````
```suggestion
Note that all the suggester functions are registered in `@misc`.
```
